server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -763,21 +751,78 @@ public void onFailure(Exception e) {
                         });
 
                         IngestDocument ingestDocument = newIngestDocument(indexRequest);
-                        executePipelines(pipelines.iterator(), hasFinalPipeline, indexRequest, ingestDocument, documentListener);
-
+                        LinkedHashSet<String> indexRecursionDetection = new LinkedHashSet<>();
+                        indexRecursionDetection.add(indexRequest.index());
+                        executePipelines(
+                            pipelines.iterator(),
+                            pipelines.hasFinalPipeline(),
+                            indexRequest,
+                            ingestDocument,
+                            documentListener,
+                            indexRecursionDetection
+                        );
                         i++;
                     }
                 }
             }
         });
     }
 
+    private Pipelines getPipelines(IndexRequest indexRequest) {
+        final String pipelineId = indexRequest.getPipeline();
+        indexRequest.setPipeline(NOOP_PIPELINE_NAME);
+        final String finalPipelineId = indexRequest.getFinalPipeline();
+        indexRequest.setFinalPipeline(NOOP_PIPELINE_NAME);
+        return new Pipelines(pipelineId, finalPipelineId);
+    }
+
+    private static class Pipelines implements Iterable<String> {
```

Comment:
I think this is a reasonable abstraction, but I don't like that we make it mutable, and that we don't encapsulate enough here.

The `withoutDefaultPipeline` method makes me uncomfortable as its name makes it sound like it would return a new object rather than making a mutable change.

I also think we don't need to have `executePipelines(...)` take a boolean, we're treating this internally as though we'll always have a list of pipelines, but we could probably get away with passing a proper object instead `Iterator<String>`. I think it'd be much clearer that way.

Reply:
I abstracted the `Iterator<String>` and the boolean flag into a `PipelineIterator`, however it's not immutable. As that's the nature of iterators, I think that's fine.

The `executePipelines` needs to know about three properties of the current pipeline: the name (to add the name in the exception in case the pipeline itself can't be resolved), the pipeline itself, and whether the current pipeline is the final pipeline (if true, it's disallowed to override _index).

The `PipelineIterator` encapsulates that state and I think that makes it cleaner than before. Thanks for the suggestion 👍